File: Jian/week8/n2579/stairs.py
# 2579번 계단 오르기
# 1. 점화식 수정
# 2. 작은 수를 넣었을 때 dp와 stairs 생성이 불가, 런타임 에러가 나서 수정

n = int(input())
dp = [0] * n
# stairs = []
# for i in range(0, n):
#     stairs.append(int(input()))
stairs = [int(input()) for _ in range(n)]

if n <= 2:
    print(sum(stairs))
else:
    dp[0] = stairs[0]
    dp[1] = stairs[1] + stairs[0]
    for i in range(2, n):
        # dp[i] = stairs[i] + max(dp[i - 1], dp[i - 2])
        dp[i] = stairs[i] + max((dp[i-3] + stairs[i-1]), dp[i-2])
    print(dp[-1])

File: Movie/W8/Stairs.java
package W8;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Stairs {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        List<Integer> list = new ArrayList<>();
        list.add(0);
        for (int i = 0; i < N; i++) {
            list.add(Integer.parseInt(br.readLine()));
        }

        int[] dp = new int[N + 1];
        dp[1] = list.get(1);
        if(N>1){
            dp[2] = list.get(2) + list.get(1);


            for (int i = 3; i <= N; i++) {
                dp[i] = Math.max(list.get(i - 1) + dp[i - 3], dp[i - 2]) + list.get(i);

            }
        }

        System.out.println(dp[N]);


    }

}
